### PR TITLE
EOS-19093: Addition of Persistent Cache for Sensor Alerts (mainly node data)

### PR DESCRIPTION
**Pull request by sumedhak27**
## Problem Statement
<pre>
  <code>
    Story Ref (if any):
    Please add Problem statement here...
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Please add Problem decription here...
  </code>
</pre>
## Solution
<pre>
  <code>
    Please add summary of the change,List any dependencies that are required for this change.
  </code>
</pre>
## Sanity testing on RPM done
<pre>
  <code>
    - [ ] Yes
    - [ ] No
  </code>
</pre>
## Unit/Manual Testing Description
<pre>
  <code>
    Please describe the tests that you ran to verify your changes.
  </code>
</pre>

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 1]
```diff
@@ -103,15 +107,22 @@ def initialize(self, conf_reader, msgQlist, product):
...
+        self._next_scheduled_time = self._polling_interval
+        self._fault_accepted_time = int(self._conf_reader._get_value_with_default(
+                                self.RAIDIntegritySensor, self.FAULT_ACCEPTED_TIME, self.DEFAULT_FAULT_ACCEPTED_TIME))
+
+        if self._polling_interval < self.MIN_POLLING_INTERVAL:
+            self._polling_interval = self.MIN_POLLING_INTERVAL
+
         self.utility = Utility()
         if self.utility.is_env_vm():
             self.shutdown()
```
Comment (sumedhak27) [inline low-level/sensors/impl/generic/raid_integrity_data.py:124]: I think this can be removed, since `sspl_constansts.py` already has a list
`DISABLE_FOR_VIRTUAL_SERVER = ("CPUFaultSensor", "RAIDIntegritySensor",  "RAIDsensor", "MemFaultSensor", "NodeHWsensor")`
which prevents these sensors to run In virtual environemnt.

File: low-level/sensors/impl/generic/raid.py [hunk 2]
```diff
@@ -132,6 +140,40 @@ def initialize(self, conf_reader, msgQlist, product):
...
+                '_devices' : self._devices,
+                '_missing_drv' : self._missing_drv,
+                '_prev_drive_dict' : self._prev_drive_dict,
+                'prev_alert_type' : self.prev_alert_type,
+            }
+            store.put(self.persistent_raid_data, self.RAID_SENSOR_DATA_PATH)
+
         # Allow systemd to process all the drives so we can map device name to serial numbers
         #time.sleep(120)
         self.utility = Utility()
```
Comment (sumedhak27) [inline low-level/sensors/impl/generic/raid.py:179]: I think this can be removed, since `sspl_constansts.py` already has a list
`DISABLE_FOR_VIRTUAL_SERVER = ("CPUFaultSensor", "RAIDIntegritySensor", "RAIDsensor", "MemFaultSensor", "NodeHWsensor")`
which prevents these sensors to run In virtual environemnt.

File: low-level/framework/platforms/realstor/realstor_enclosure.py [hunk 3]
```diff
@@ -97,6 +97,7 @@ class RealStorEnclosure(StorageEnclosure):
 
     poll_system_ts = 0
     mc_timeout_counter = 0
+    ws_responce_status = None
```
Comment (cdeshmukh) [inline low-level/framework/platforms/realstor/realstor_enclosure.py:100]: spell check - response

Also i dont see the need for this

File: low-level/sensors/impl/platforms/realstor/realstor_enclosure_sensor.py [hunk 4]
```diff
@@ -131,7 +150,8 @@ def run(self):
 
                 self.fault_alert = True
 
-            elif mc_timeout_counter == 0 and self.previous_alert_type != self.rssencl.FRU_FAULT_RESOLVED \
+            elif mc_timeout_counter == 0 and  ws_response_status == 200 \
```
Comment (cdeshmukh) [inline low-level/sensors/impl/platforms/realstor/realstor_enclosure_sensor.py:153]: what is this logic change, why necessary? Why we are touching unrelated code?

@Sumedh-a-k  in general, dont increase scope of the change beyond whats discussed, it increases test scope too.

File: low-level/message_handlers/node_data_msg_handler.py [hunk 5]
```diff
@@ -195,6 +195,59 @@ def initialize(self, conf_reader, msgQlist, product):
...
+        if os.path.isfile(self.DISK_SENSOR_DATA_PATH):
+            self.persistent_disk_sensor_data = store.get(self.DISK_SENSOR_DATA_PATH)
+        if self.persistent_disk_sensor_data:
+            if self.persistent_disk_sensor_data['disk_fault'] == "True":
+                self.disk_fault = True
+            else:
+                self.disk_fault = False
+        else:
+            self.persistent_disk_sensor_data = {
+                'disk_fault' : str(self.disk_fault),
```
Comment (cdeshmukh) [inline low-level/message_handlers/node_data_msg_handler.py:229]: What is 'self.disk_fault' before this, NULL? Why not set to false here?

File: low-level/sensors/impl/generic/raid.py (hunk 2 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid.py:179]: no, these are self sensor measures, when someone adds up in config by mistake or without enough knowledge.
again, dont touch or change anything beyond decided scope of any task, it becomes difficult to understand, explain the reason and test scope increases unecessarily.

File: low-level/sensors/impl/generic/raid.py [hunk 6]
```diff
@@ -282,7 +324,7 @@ def _process_mdstat(self):
             # Parse out status' and path info for each drive
             if "md" in fields[0]:
                 self._device = f"/dev/{fields[0]}"
-                self._devices.append(self._device)
+                self._devices.add(self._device)
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/raid.py:327]: we need to maintain order of devices as it appear in conf file. why change here?

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 7]
```diff
@@ -59,9 +59,13 @@ class RAIDIntegritySensor(SensorThread, InternalMsgQ):
     RACK_ID = "rack_id"
     POLLING_INTERVAL = "polling_interval"
     TIMESTAMP_FILE_PATH_KEY = "timestamp_file_path"
+    FAULT_ACCEPTED_TIME = "fault_accepted_time"
 
-    # check once a week (below time is in seconds), the integrity of raid data
-    DEFAULT_POLLING_INTERVAL = "604800"
+    # check once in two weeks (below time is in seconds), the integrity of raid data
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/raid_integrity_data.py:64]: Update comment
\# Scan for RAID integrity error every 2 weeks (1209600 seconds)

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 8]
```diff
@@ -59,9 +59,13 @@ class RAIDIntegritySensor(SensorThread, InternalMsgQ):
...
+    DEFAULT_POLLING_INTERVAL = "1209600"
+    # interval to check integrity can not be set below one day i.e. 86400 seconds.
+    MIN_POLLING_INTERVAL = 86400
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/raid_integrity_data.py:67]: \ # Minimum allowed frequency for RAID integrity scans is 1 day (86400 seconds ),
\ as frequent scans affect disk i/o performance

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 1 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid_integrity_data.py:124]: no, check similar previous comment

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 9]
```diff
@@ -103,15 +107,22 @@ def initialize(self, conf_reader, msgQlist, product):
         self._rack_id = self._conf_reader._get_value_with_default(
                                 self.SYSTEM_INFORMATION, COMMON_CONFIGS.get(self.SYSTEM_INFORMATION).get(self.RACK_ID), '001')
         self._node_id = self._conf_reader._get_value_with_default(
-                                self.SYSTEM_INFORMATION, COMMON_CONFIGS.get(self.SYSTEM_INFORMATION).get(self.NODE_ID), '001')  
+                                self.SYSTEM_INFORMATION, COMMON_CONFIGS.get(self.SYSTEM_INFORMATION).get(self.NODE_ID), '001')
         self._timestamp_file_path = self._conf_reader._get_value_with_default(
                                     self.RAIDIntegritySensor, self.TIMESTAMP_FILE_PATH_KEY, self.DEFAULT_TIMESTAMP_FILE_PATH)
         self._polling_interval = int(self._conf_reader._get_value_with_default(
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/raid_integrity_data.py:113]: rename '_polling_interval ' to '_scan_frequency'

**Review by cdeshmukh — COMMENTED** [review]
Limit changes to what is required only, dont change beyond scope as it increases test scope too, which is not intended for cortx-1.0 branch

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 [hunk 10]
```diff
@@ -104,6 +104,11 @@ max_login_attempts=2
 threaded=true
 RAID_status_file=/proc/mdstat
 
+[RAIDINTEGRITYSENSOR]
+polling_interval=1209600
+timestamp_file_path=/var/cortx/sspl/data/raid_integrity/last_execution_time
+fault_accepted_time=86400
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:110]: once fault has occured after `fault_accepted_time` we will send second fault message to contact the Seagate Support, that I kept configurable.

File: low-level/framework/platforms/realstor/realstor_enclosure.py (hunk 3 above)
Comment (sumedhak27) [reply low-level/framework/platforms/realstor/realstor_enclosure.py:100]: just the `mc_timeout_counter` is not enough to decide for fault_resolved alert as we are fetching whether fault was previously present and previous fault type from persistent cache, in next run cycle all 3 parameters, mc_timeout_count=0, falut_alert = True, and previous_fault != FAULT_RESOLVED will be true, so even tough connection has failed it will give FAULT_RESOLVED, when fault is not resolved, thus response_status is required.

File: low-level/message_handlers/node_data_msg_handler.py [hunk 11]
```diff
@@ -195,6 +195,44 @@ def initialize(self, conf_reader, msgQlist, product):
...
+
+        # Persistent Cache for CPU sensor  
+        self.CPU_SENSOR_DATA_PATH = os.path.join(cache_dir_path, f'CPU_SENSOR_DATA_{self.node_id}')
+        
+        self.persistent_cpu_sensor_data = store.get(self.CPU_SENSOR_DATA_PATH)
+        if self.persistent_cpu_sensor_data:
+            self.cpu_fault = bool(self.persistent_cpu_sensor_data['cpu_fault'])
+        else:
+            self.persistent_cpu_sensor_data = {
+                'cpu_fault' : str(self.cpu_fault),
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:212]: already initialized along with ohter class variables.

File: low-level/message_handlers/node_data_msg_handler.py (hunk 5 above)
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:229]: Already initialized to false along with the class member.

File: low-level/message_handlers/node_data_msg_handler.py [hunk 12]
```diff
@@ -517,48 +562,51 @@ def _generate_cpu_data(self):
...
-        if self._node_sensor.cpu_usage >= self._cpu_usage_threshold:
+        if self._node_sensor.cpu_usage >= self._cpu_usage_threshold \
+            and not self.cpu_fault:
 
-            if not self.cpu_fault :
-                self.cpu_fault = True
-                # Create the cpu usage data message and hand it over to the egress processor to transmit
-                fault_event = "CPU usage increased to %s, beyond configured threshold of %s" \
-                                %(self._node_sensor.cpu_usage, self._cpu_usage_threshold)
+            self.cpu_fault = True
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:583]: 1. high_cpu_usage
2. high_disk_space_usage
3. high_memory_usage
is this okay??

File: low-level/message_handlers/node_data_msg_handler.py [hunk 13]
```diff
@@ -517,48 +562,51 @@ def _generate_cpu_data(self):
...
-                self.os_sensor_type["cpu_usage"] = self.cpu_sensor_data
-                # Transmit it out over rabbitMQ channel
-                self._write_internal_msgQ(RabbitMQegressProcessor.name(), jsonMsg)
+            # Add in uuid if it was present in the json request
+            if self._uuid is not None:
+                cpuDataMsg.set_uuid(self._uuid)
+            jsonMsg = cpuDataMsg.getJson()
+            self.cpu_sensor_data = jsonMsg
+            self.os_sensor_type["cpu_usage"] = self.cpu_sensor_data
+            # Transmit it out over rabbitMQ channel
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:618]: This is cortx-1.0 branch. Rabbitmq is needed here.

File: low-level/sensors/impl/generic/raid.py (hunk 2 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid.py:179]: okay..

File: low-level/sensors/impl/generic/raid.py (hunk 6 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid.py:327]: the same device used to get added in every iteration. set is efficient in search that's why I used but if we need to maintain order then I can just check once if a device is present in the list or not before adding.

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 9 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py:113]: even in sspl conf??

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 1 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py:124]: sure.

File: low-level/sensors/impl/platforms/realstor/realstor_enclosure_sensor.py (hunk 4 above)
Comment (sumedhak27) [reply low-level/sensors/impl/platforms/realstor/realstor_enclosure_sensor.py:153]: sure, as explained in the previous comment, due to persistent cache addition, we need this variable.

File: low-level/message_handlers/node_data_msg_handler.py [hunk 14]
```diff
@@ -195,6 +195,44 @@ def initialize(self, conf_reader, msgQlist, product):
 
         self._import_products(product)
         cache_dir_path = os.path.join(DATA_PATH, self.CACHE_DIR_NAME)
+
+        # Persistent Cache for CPU sensor  
+        self.CPU_SENSOR_DATA_PATH = os.path.join(cache_dir_path, f'CPU_SENSOR_DATA_{self.node_id}')
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:200]: should I change host_sensor to memory_sensor in every variable then?? as you asked to change host_fault -> high_memory_usage_fault

[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:200]: then names can be, 
CPU_FAULT_SENSOR_DATA_PATH
DISK_FAULT_SENSOR_DATA_PATH
MEMORY_FAULT_SENSOR_DATA_PATH    OR    HOST_FAULT_SENSOR_DATA_PATH

persistent_cpu_fault_sensor_data
persistent_disk_fault_sensor_data
persistent_memory_fault_sensor_data   OR  persistent_host_fault_sensor_data

[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py:200]: I will need to change nw fault names also then??
NW_FAULT_SENSOR_DATA_PATH
persistent_nw_fault_sensor_data
These would be long names, can we shorten them??

File: low-level/sensors/impl/generic/raid.py (hunk 6 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid.py:327]: no checking if device would not main order from conf file. we just need to clear _devices list during each iteration then?
is that okay?

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 [hunk 15]
```diff
@@ -104,6 +104,11 @@ max_login_attempts=2
 threaded=true
 RAID_status_file=/proc/mdstat
 
+[RAIDINTEGRITYSENSOR]
+polling_interval=1209600
+timestamp_file_path=/var/cortx/sspl/data/raid_integrity/last_execution_time
```
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:109]: yes it's already there in sspl constants.

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 16]
```diff
@@ -103,15 +107,22 @@ def initialize(self, conf_reader, msgQlist, product):
...
                                 self.RAIDIntegritySensor, self.POLLING_INTERVAL, self.DEFAULT_POLLING_INTERVAL))
+        self._next_scheduled_time = self._polling_interval
+        self._fault_accepted_time = int(self._conf_reader._get_value_with_default(
```
Comment (shriya-deshmukh) [inline low-level/sensors/impl/generic/raid_integrity_data.py:116]: Please make sure to read this newly added key-value from /etc/sspl.conf file, 
because it will not be available in consul  as we didn't added it in sspl.sls file.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 15 above)
[the remark this answers is not in the document]
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:109]: Then why its still seen here in config. Confg holds entries which needs to be configurable, doesnt apply for 'timestamp_file_path'

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 10 above)
[the remark this answers is not in the document]
Comment (cdeshmukh) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:110]: We never decided on second fault. In first fault itself, had asked for a description stating "if fault persists for more than 1-2 days, contact seagate support " !!

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 7 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid_integrity_data.py:64]: comment not addressed

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 8 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid_integrity_data.py:67]: comment not addressed

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 9 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid_integrity_data.py:113]: yes

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 16 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid_integrity_data.py:116]: oh good catch @shriya-deshmukh , that case add a seperate PR for cortx-provisioner too using cortx-1.0 branch, to update sspl.sls

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 10 above)
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1:110]: Ohh so the first fault alert only needs to have a description change,  no logic is required then for the second fault after 1 or 2 days?

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 16 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py:116]: won't be required if we don't want to send 2nd alert after 1.2 days of fault in raid_integrity sensor.
this variable will be removed.

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 9 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py:113]: here if we change it to `scan_frequency` instead of `polling_interval` in conf then we need to update sspl.sls.

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 8 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py]: addressed.

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 7 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py]: addressed.

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 15 above)
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1]: removed.

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 16 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py]: removed.

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 9 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py]: changed to `scan_frequency` only in raid_integrity Sensor, 
for `sspl.sls` and `sspl.conf`  => `polling_interval` is kept as it was.

File: low-level/sensors/impl/generic/raid.py (hunk 6 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid.py]: cleared _devices list in each iteration and kept is as list preserving order from /proc/mdstat file.

File: low-level/message_handlers/node_data_msg_handler.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py]: implemented. 👍

File: low-level/message_handlers/node_data_msg_handler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/message_handlers/node_data_msg_handler.py]: variable names changed 👍

File: low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1 (hunk 10 above)
[the remark this answers is not in the document]
Comment (sumedhak27) [reply low-level/files/opt/seagate/sspl/conf/sspl.conf.LDR_R1]: Logic of 2nd alert removed. added to the description of first alert.

File: low-level/message_handlers/node_data_msg_handler.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (cdeshmukh) [reply low-level/message_handlers/node_data_msg_handler.py:619]: yes sorry, i mixed up

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 17]
```diff
@@ -195,14 +201,18 @@ def _raid_health_monitor(self):
                             data = fs.read().rstrip()
                         if self.FAULT_RESOLVED in data:
                             self.alert_type = self.FAULT
-                            self._alert_msg = "RAID disks present in %s RAID array, needs synchronization." %device
+                            self._alert_msg = "RAID disks present in %s RAID array, needs synchronization." %device +\
+                                " If fault persists for more than 1 or 2 days, Please contact Seagate support."
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/raid_integrity_data.py:205]: just change this to
 If fault persists for more than 2 days, Please contact Seagate support

File: low-level/sensors/impl/generic/raid_integrity_data.py [hunk 18]
```diff
@@ -195,14 +201,18 @@ def _raid_health_monitor(self):
...
                             self._send_json_msg(self.alert_type, device, self._alert_msg)
                             self._update_fault_state_file(device, self.FAULT, fault_status_file)
+                            self._scan_frequency = self.MIN_SCAN_FREQUENCY
                     else:
                         self.alert_type = self.FAULT
-                        self._alert_msg = "RAID disks present in %s RAID array, needs synchronization." %device
+                        self._alert_msg = "RAID disks present in %s RAID array, needs synchronization." %device +\
+                                " If fault persists for more than 1 or 2 days, Please contact Seagate support."
```
Comment (cdeshmukh) [inline low-level/sensors/impl/generic/raid_integrity_data.py:212]: same here
 If fault persists for more than 2 days, Please contact Seagate support

Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py]: done 👍

File: low-level/sensors/impl/generic/raid_integrity_data.py (hunk 17 above)
Comment (sumedhak27) [reply low-level/sensors/impl/generic/raid_integrity_data.py]: done.

File: low-level/sensors/impl/generic/raid.py (hunk 6 above)
Comment (cdeshmukh) [reply low-level/sensors/impl/generic/raid.py]: understood, thanks

**Review by cdeshmukh — APPROVED** [review]

**Review by shriya-deshmukh — APPROVED** [review]
